### ISSUE TITLE
Enable Bulk Partition Drops using Hive MS 

### ISSUE THREAD

**Issue by ramyarajasekaran** [trinodb/trino#7249]
Hive Metastore offers a method to drop multiple partitions at the same time. Currently, partitions are dropped individually and this takes too long for tables with many partitions (>10K). Would be awesome if Trino allowed bulk partition drops.

Hive Issue to add bulk partition drops 
[https://issues.apache.org/jira/browse/HIVE-6256](https://issues.apache.org/jira/browse/HIVE-6256)
HMS API for bulk partition drop [here](https://hive.apache.org/javadocs/r2.1.1/api/org/apache/hadoop/hive/metastore/IMetaStoreClient.html#dropPartitions-java.lang.String-java.lang.String-java.util.List-boolean-boolean-)

**Comment by findepi** [comment]
@electrum @phd3 @losipiuk @djsstarburst what do you think?

**Comment by phd3** [comment]
It makes sense to me to try out and verify its behavior for different hive versions. looks like in Trino, we'll need to use the `ThriftMetastore.IFace#drop_partitions_req` call.

@ramyarajasekaran Currently, `hive.max-concurrent-metastore-drops` property can be tuned to improve parallelism for such drops. However, IIRC there's still a [race condition](https://github.com/trinodb/trino/issues/3898#issuecomment-637110083) possibility especially for nested partitions. It'll be good to see if using the bulk API circumvents that as well.

**Comment by ChinmaySKulkarni** [comment]
@findepi @phd3 @electrum @losipiuk @djsstarburst Seems like if we were to use the `ThriftMetastore.IFace#drop_partitions_req` call, we have to serialize the drop partition expression in our `DropPartitionsExpr` and potentially implement Hive's [PartitionExpressionProxy](https://github.com/apache/hive/blob/a93798f40a81506f2b0fdee224465d432161a0e0/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/PartitionExpressionProxy.java) which can allow us to define how the [partExprs](https://hive.apache.org/javadocs/r2.1.1/api/org/apache/hadoop/hive/metastore/IMetaStoreClient.html#dropPartitions-java.lang.String-java.lang.String-java.util.List-boolean-boolean-) `byte[]` is evaluated. 

The default implementation ([DefaultPartitionExpressionProxy](https://github.com/apache/hive/blob/a93798f40a81506f2b0fdee224465d432161a0e0/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/DefaultPartitionExpressionProxy.java)) which is what we set as the value for `metastore.expression.proxy` when running standalone HMS with Trino (see [this](https://cwiki.apache.org/confluence/display/Hive/AdminManual+Metastore+3.0+Administration#AdminManualMetastore3.0Administration-RunningtheMetastoreWithoutHive)) just throws `UnsupportedOperationException` for all methods.

This would likely require a patch to Hive Metastore. Are we okay with having a contribution that depends on a specific Hive patch?

**Comment by findepi** [comment]
> This would likely require a patch to Hive Metastore. 

Would that mean it's hardly usable by the general audience?

I would strongly prefer that we find something that can work for everyone, or at least for large group of users.
So for example depending on some reasonably new (existing in the wild) Hive version is OK for me, even if some people still use older versions.

**Comment by ChinmaySKulkarni** [comment]
@findepi I agree that this should work for as many users as possible, if not all of them. Based on my initial poking around with Hive Metastore, it looks like there isn't a way to use the batch drop partitions API without the above-mentioned changes to HMS **OR** we could drop a jar containing this new class (our new implementation of PartitionExpressionProxy) which we will have to make available to the HMS classpath. 

The latter approach requires some manual changes for users and the former, like you mentioned would make it such that we can't use this in Trino until the next Hive release. Any contributors with more experience making Hive metastore changes that you are aware of and can add to this discussion?

FWIW, I have brought this issue up in the Hive community with [this dev-list email](https://lists.apache.org/thread.html/r5367e1596271b3eab3ce6504d880b4e20bbf509fefcaa8f0ec86ed2c%40%3Cdev.hive.apache.org%3E) and found a similar Hive Jira ([HIVE-23851](https://issues.apache.org/jira/browse/HIVE-23851?focusedCommentId=17388228&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17388228)) where they had to implement their own PartitionExpressionProxy implementation to support msck and hence might be similar to the approach we need to take.

**Comment by findepi** [comment]
@ChinmaySKulkarni 
maybe we should offer a patch to Hive project which would add support for bulk partition drop to HMS (out of the box).
then we could offer a backport of such a change to Hive 3.x and 2.x branches. If Hive project is OK with that, that would mean the feature can be made available to HMS users is reasonable timeframe.

**Comment by ChinmaySKulkarni** [comment]
@findepi sorry for the delay in circling back here. I have created [HIVE-25463](https://issues.apache.org/jira/browse/HIVE-25463) to gauge the Hive community's feedback regarding the same. Hopefully if they are ok with this approach, I will contribute a patch and see where we can go from there. Thanks.
